src/cyclic_queue.hpp

```diff
@@ -0,0 +1,165 @@
+#ifndef CYCLIC_QUEUE_HPP
+#define CYCLIC_QUEUE_HPP
+
+#include <initializer_list>
+
+/**
+ * @brief Cyclic Queue
+ *
+ * @details
+ * This container holds values in a C-style array.
+ * The array loops back on itself to allow a cyclic pattern.
+ * The array can be accessed as a queue but can also pop the latest addition out.
+ */
+template <typename T, int size>
+class CyclicQueue {
+  private:
+    T *start, *end;
+    T *head, *tail;
+    T values[size];
+    int count = 0;
+    /**
+     * @brief Increment tail pointer
+     *
+     * @description
+     * Increments the tail pointer or loops it around. If the head is replaced, increment head.
+     */
+    void incrementTail() {
+        if (tail == end) {
+            tail = start;
+        } else {
+            tail++;
+        }
+        count++;
+        if (tail == head && count > 1) {
+            incrementHead();
+        }
+    }
+    /**
+     * @brief Increment head pointer
+     *
+     * @description
+     * Increments the tail pointer or loops it around.
+     */
+    void incrementHead() {
+        if (head == end) {
+            head = start;
+        } else {
+            head++;
+        }
+        count--;
+    }
+    /**
+     * @brief Decrement tail pointer
+     *
+     * @description
+     * Decrements the tail pointer or loops it around.
+     */
+    void decrementTail() {
+        if (tail == start) {
+            tail = end;
+        } else {
+            tail--;
+        }
+        count--;
+    }
+
+  public:
+    CyclicQueue(std::initializer_list<T> init) : start(values), end(values + size - 1), head(start), tail(end) {
+        for (int value : init) {
```

Comment:
This causes the class to only be able to handle intergers, thus making the template functionality useless